### DISCUSSION TITLE
Applying serializer updates to model without committing?

### DISCUSSION THREAD

**Discussion by Kangaroux** [encode/django-rest-framework#8570] [Question & Answer] [answered]
Our application requires changes to certain models to be approved by users. We'd like to be able to display an accurate representation of the updates to the user. We can't just display the raw `data` dict, because there may be properties or other fields impacted by the change.

Ideally, we'd apply the change to the instance without committing it to the database. This way we can serialize it and get a true preview of the changes.

A couple hacky solutions that came to mind:
1. Use a transaction to save the model and serialize it, then rollback the transaction.
2. Temporarily replace the `.save()` method on the model instance with a dummy method, so that when the serializer calls `instance.save()` nothing happens. Because `.save()` might make some changes, I'm less fond of this method

**Comment by Kangaroux** [comment, marked as answer]
I ended up going with the transaction rollback solution and it worked out surprisingly well.

1. Store the original object and make a copy of it
2. Modify the copy in a transaction, then raise an exception to rollback the transaction
3. Restore the original object
4. You now have both the original (restored) object and the modified one.

Code example:

```python
from copy import copy
from django.db import transaction

class ExitTransaction(Exception):
    pass

# ...

# Make a copy of the original so we can restore it later. Rolling back a
# transaction doesn't undo the changes to the instance.
original = instance.obj
instance.obj = copy(original)

try:
    with transaction.atomic():
        # Modify instance.obj somehow...
        modify_obj(instance.obj)

        # Raise an exception to rollback the transaction
        raise ExitTransaction 
except ExitTransaction:
    pass

# instance.obj still has the modifications
updated_obj = instance.obj

# Restore the original object
instance.obj = original

# updated_obj can now be used however you need it, but just make sure not to
# call .save() or else it will actually commit the changes.
```

**Reply by voxvici** [reply]
thanks for sharing solution, interesting approach
